src/audio/codec_adapter/codec/generic.c

```diff
@@ -0,0 +1,366 @@
+	struct list_item *_mem_list;
+
+	if (!ptr) {
+		comp_err(dev, "codec_free_memory: error: NULL pointer passed.");
+		return -EINVAL;
+	}
+	/* Find which container keeps this memory */
+	list_for_item_safe(mem_list, _mem_list, &cd->codec.memory.mem_list) {
+		mem = container_of(mem_list, struct codec_memory, mem_list);
+		if (mem->ptr == ptr) {
+			rfree(mem->ptr);
+			list_item_del(&mem->mem_list);
+			rfree(mem);
+			return 0;
+		}
+	}
+
+	comp_err(dev, "codec_free_memory: error: could not find memory pointed by %p",
+		 (uint32_t)ptr);
+
+	return -EINVAL;
+}
+
+static int validate_config(struct codec_config *cfg)
+{
+	//TODO: validation of codec specific setup config
+	return 0;
+}
+
+int codec_prepare(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t codec_id = cd->ca_config.codec_id;
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_prepare() start");
+
+	if (cd->codec.state == CODEC_PREPARED)
+		return 0;
+	if (cd->codec.state < CODEC_INITIALIZED)
+		return -EPERM;
+
+	ret = codec->ops->prepare(dev);
+	if (ret) {
+		comp_err(dev, "codec_prepare() error %d: codec specific prepare failed, codec_id 0x%x",
+			 ret, codec_id);
+		goto end;
+	}
+
+	codec->s_cfg.avail = false;
+	codec->r_cfg.avail = false;
+	codec->r_cfg.data = NULL;
+	list_init(&codec->memory.mem_list);
+
+	if (codec->r_cfg.data)
+		rfree(codec->r_cfg.data);
+
+	codec->state = CODEC_PREPARED;
+	comp_dbg(dev, "codec_prepare() done");
+end:
+	return ret;
+}
+
+int codec_process(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t codec_id = cd->ca_config.codec_id;
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_process() start");
+
+	if (cd->codec.state < CODEC_PREPARED) {
+		comp_err(dev, "codec_prepare(): wrong state of codec %x, state %d",
+			 cd->ca_config.codec_id, codec->state);
+		return -EPERM;
+	}
+
+	ret = codec->ops->process(dev);
+	if (ret) {
+		comp_err(dev, "codec_prepare() error %d: codec process failed for codec_id %x",
+			 ret, codec_id);
+		goto out;
+	}
+
+	comp_dbg(dev, "codec_process() done");
+out:
+	return ret;
+}
+
+int codec_apply_runtime_config(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t codec_id = cd->ca_config.codec_id;
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_apply_config() start");
+
+	if (cd->codec.state < CODEC_PREPARED)
+		return -EPERM;
+
+	if (cd->codec.state < CODEC_PREPARED) {
+		comp_err(dev, "codec_prepare() wrong state of codec %x, state %d",
+			 cd->ca_config.codec_id, codec->state);
+		return -EPERM;
+	}
+
+	ret = codec->ops->apply_config(dev);
+	if (ret) {
+		comp_err(dev, "codec_apply_config() error %d: codec process failed for codec_id %x",
+			 ret, codec_id);
+		goto out;
+	}
+
+	cd->codec.r_cfg.avail = false;
+	rfree(codec->r_cfg.data);
+	codec->r_cfg.data = NULL;
+
+	comp_dbg(dev, "codec_apply_config() end");
```

Comment:
Dont need the end debug

Reply:
Are you sure? If there is some problem with that (investigation of some BUG) its good to have both starting and ending logs - if the ending one is missing you know sth wrong happened in between. Do you agree?

Reply:
Insert ending logs whenever you are actually debugging. Normal runtime it's too much spam.

Reply:
@paulstelian97, hence `comp_dbg()` - you enable when needed.

Reply:
Still atypical but not enough for me to block.

Reply:
The only place I would expect to see ```dbg("func enter");``` and ```dbg("func leave");``` is around any API call into 3rd party binary code. i.e. untrusted code.

Reply:
@lgirdwood correct and this is exactly what happens here -> `ret = codec->ops->apply_config(dev);` does call 3rd party binary API to apply that config by **`API_CALL(cd, XA_API_CMD_SET_CONFIG_PARAM, param->id, param->data, ret);`**

Does it convince you or let's just remove that end comment?

Reply:
I would put a comment around any applicable start/end debug stating this is around an "untrusted binary code" so it's clear for anyone debugging.